README.md
s3000_ros
=========

Ros Hydro compatible (Catkinized) version of the S3000 Lidar ROS package from Robotnik Automation (GPLv3)

License
=======

Licensed under GPLv3. Please checkout the `gitlab-devel` branch for the most recent mirror of our internal changes.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package s3000_laser
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.8.2 (2018-10-23)
------------------
* fixes for narrowing issues with c++11
* Contributors: Tracey Spicer

0.8.1 (2017-12-07)
------------------
* CORE-9158: Better interrupt byte check.
* Contributors: Mike Purvis

0.8.0 (2017-11-29)
------------------
* CORE-8925: Try harder to reset laser after detecting error.
* Contributors: Mike Purvis

0.7.2 (2017-05-03)
------------------
* CORE-7222: Recover gracefully from bad TX comms
* CORE-6899: Re-poll code when H or G.
* Contributors: Mike Purvis

0.7.1 (2017-03-10)
------------------
* CORE-6843: Fix SEG_N value, now can properly recover N1 errors.
* Add missing boost/scoped_ptr include.
* Contributors: Mike Purvis

0.7.0 (2017-03-01)
------------------
* CORE-6646: Error counters
* CORE-6647, CORE-6645: Updates to fault resetting.
* CORE-6505, CORE-6506: Re-enable error diagnostics, with error resets.
* Contributors: Mike Purvis

0.6.0 (2016-12-16)
------------------
* Better distinguish unset from zero.
* Report SCID values as hex.
* Bump version number.
* Contributors: Mike Purvis

0.1.7 (2016-11-14)
------------------
* Add missing s3000_laser to install targets.
* Contributors: Justin Godson

0.1.6 (2016-11-14)
------------------
* Add block check call on rostopic to s3000_laser.cc
* Split ReadLaser and move read delay out of driver.
* Add block diagnostics check functionality to s3000 driver.
* Contributors: Justin Godson

0.1.5 (2016-05-09)
------------------
* Parameterize frequency warning tolerance, change default from 5% to 20%.
* Contributors: Mike Purvis

0.1.4 (2016-02-16)
------------------
* Set range max from param, better default value.
* Contributors: Enrique Fernandez

0.1.3 (2016-02-03)
------------------
* Reduce log outputs from error states.
* Contributors: Mike Purvis

0.1.2 (2015-12-22)
------------------
* Connection/reconnection fix.
  This catches another serial failure mode, this time where select()
  reports the device as ready to read, but no actual data is returned.
  This previously manifested as simply being stuck in an unproductive
  read loop, but now it correctly recovers from this state.
* Contributors: Mike Purvis

0.1.1 (2015-12-18)
------------------
* Eliminate fixed rate loop, block on data receipt
* Rename connection diagnostic, remove self_test.
* Remove getData, use BlockOnRead.
* Switch to DiagnosedPublisher, scoped_ptr.
* Remove tf broadcaster, doesn't belong here.
* Package format 2, CMakeLists cleanup, maintainer change.
* Contributors: Mike Purvis

0.1.0 (2015-09-21)
------------------
* Merge pull request `#2 <https://github.com/clearpathrobotics/s3000_laser/issues/2>`_ from clearpathrobotics/igrrs-441-fix
  Reverted retrieval of params and fixed launch file
* Updated launch file
* Reverted retrieval of params and fixed launch file
  Fixed port param
* Merge pull request `#1 <https://github.com/clearpathrobotics/s3000_laser/issues/1>`_ from clearpathrobotics/igrrs-441
  IGRRS-441: Fix LIDAR diagnostics to return sane results
* Fixed LIDAR status and frequency diagnostics
  Added check for no data from laser
  Updated diagnostics
  Minor refactoring
  Fixed launch file param and minor formatting
  Formatting + removed whitespace
  Updated diagnostic status to use enums
  Formatting
* Added serial device library and launch file to install targets
* Added proper install target in CMakeLists.txt
* remove ~ files
* Changed back to s3000_laser
* First Commit of Catkinized S3000 driver for ROS
* Initial commit
* Contributors: Jonathan Jekir, Mustafa Safri, Prasenjit Mukherjee, abencz, ibaranov, ibaranov-cp, rocksteady_robot


test/stats.py
#!/usr/bin/env python

from re import findall

with open('laser-14.05s.log') as f:
    t = f.read()

print len(t), "bytes"
print len(findall("\0\0\0\0\0\0", t))



src/sick_s3000.cc
/*
*  Copyright (c) 2012, Robotnik Automation, SLL
*
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/
/*
 Desc: Driver for the SICK S3000 laser
 Author: Robotnik Automation SLL (based on sicks3000 by Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard for Player/Stage)
 Date: 1 Sept 2012

 The sicks3000 driver controls the SICK S 3000 safety laser scanner interpreting its data output.
 The driver is very basic and assumes the S3000 has already been configured to continuously output
 its measured data on the RS422 data lines.
*/

#include <netinet/in.h>  // htons
#include "s3000_laser/sick_s3000.h"
#include "serial/serial.h"

// 1 second of data at 500kbaud
#define DEFAULT_RX_BUFFER_SIZE 500*1024/8

////////////////////////////////////////////////////////////////////////////////
// Device codes

#define STX     0x02
#define ACK     0xA0
#define NACK    0x92
#define CRC16_GEN_POL 0x8005

//! Converts degrees to radians
double DTOR(double val)
{
    return val*3.141592653589793/180.0;
}

/*! \fn SickS3000::SickS3000()
 *  \brief Public constructor
*/
SickS3000::SickS3000( std::string port )
{
  rx_count = 0;
  // allocate our recieve buffer
  rx_buffer_size = DEFAULT_RX_BUFFER_SIZE;
  rx_buffer = new uint8_t[rx_buffer_size];
  assert(rx_buffer);

  recognisedScanner = false;
  mirror = 0;  // TODO move to property

  // Create serial port
  serial_ = new serial::Serial();
  serial_->setPort(port);
  serial_->setBaudrate(S3000_DEFAULT_TRANSFERRATE);

  return;
}

SickS3000::~SickS3000()
{
  Close();
  delete [] rx_buffer;
}

/*! \fn int SickS3000::Open()
 *  \brief Open serial port
 *  \returns -1 Error
 *  \returns 0 Ok
*/
int SickS3000::Open()
{
  serial_->open();

  if (!serial_->isOpen())
  {
    ROS_ERROR("SickS3000::Open: Error Opening Serial Port");
    return -1;
  }

  ROS_INFO_STREAM("SickS3000 serial port opened at " << serial_->getPort());

  return 0;
}

/*! \fn int SickS3000::Close()
 *  \brief Closes serial port
 *  \returns ERROR
 *  \returns OK
*/
int SickS3000::Close()
{
  if (serial_)
  {
    delete serial_;
  }
  return 0;
}

////////////////////////////////////////////////////////////////////////////////
// Set up scanner parameters based on number of results per scan
void SickS3000::SetScannerParams(sensor_msgs::LaserScan& scan, int data_count)
{
  if (data_count == 761) // sicks3000
  {
    // Scan configuration
    scan.angle_min  = static_cast<float> (DTOR(-95));
    scan.angle_max  = static_cast<float> (DTOR(95));
    scan.angle_increment =  static_cast<float> (DTOR(0.25));
    scan.time_increment = (1.0/16.6) / 761.0;  //(((95.0+95.0)/0.25)+1.0);    // Freq 16.6Hz / 33.3Hz
    scan.scan_time = (1.0/16.6);
    scan.range_min  =  0;
    scan.range_max  =  49;   // check ?

    recognisedScanner = true;
  }
}

  int SickS3000::ReadLaser( sensor_msgs::LaserScan& scan, bool& bValidData ) // public periodic function
  {
    uint8_t buffer[4000] = "\0";    // Max in 1 read

    // Read controller messages
    //if (serial->ReadPort(cReadBuffer, &read_bytes, 2000) < 0)
    ssize_t read_bytes = serial_->read(buffer, sizeof(buffer));
    if (read_bytes < 0)
    {
        ROS_ERROR("SickS3000::ReadLaser: Error reading port");
        return -1;
    }

    unsigned int messageOffset = rx_count;
    rx_count += read_bytes;
    if (rx_count > rx_buffer_size)
    {
       ROS_WARN("S3000 Buffer Full");
       rx_count = 0;
    }
    else
    {
       memcpy(&rx_buffer[messageOffset], buffer, read_bytes);
       ProcessLaserData(scan, bValidData);
    }

    return 0;
  }


int SickS3000::ProcessLaserData(sensor_msgs::LaserScan& scan, bool& bValidData)
{
  while(rx_count >= 22)
  {
    // find our continuous data header
    unsigned int ii;
    bool found = false;
    for (ii = 0; ii < rx_count - 22; ++ii)
    {
      if (memcmp(&rx_buffer[ii],"\0\0\0\0\0\0",6) == 0)
      {
        memmove(rx_buffer, &rx_buffer[ii], rx_count-ii);
        rx_count -= ii;
        found = true;
        break;
      }
    }

    if (!found)
    {
      memmove(rx_buffer, &rx_buffer[ii], rx_count-ii);
      rx_count -= ii;
      return 0;
    }

    // get relevant bits of the header
    // size includes all data from the data block number
    // through to the end of the packet including the checksum
    unsigned short size = 2*htons(*reinterpret_cast<unsigned short *> (&rx_buffer[6]));

    if (size > rx_buffer_size - 26)
    {
      ROS_WARN("S3000: Requested Size of data is larger than the buffer size");
      memmove(rx_buffer, &rx_buffer[1], --rx_count);
      return 0;
    }

    // check if we have enough data yet
    if (size > rx_count - 4)
      return 0;

    unsigned short packet_checksum = *reinterpret_cast<unsigned short *> (&rx_buffer[size+2]);
    unsigned short calc_checksum = CreateCRC(&rx_buffer[4], size-2);
    if (packet_checksum != calc_checksum)
    {
      ROS_WARN("S3000: Checksum's dont match, thats bad (data packet size %d)",size);
      memmove(rx_buffer, &rx_buffer[1], --rx_count);
      continue;
    }
    else
    {
      uint8_t * data = &rx_buffer[20];
      if (data[0] != data[1])
      {
        ROS_WARN("S3000: Bad type header bytes dont match");
      }
      else
      {
        if (data[0] == 0xAA)
        {
          ROS_WARN("S3000: We got a I/O data packet we dont know what to do with it\n");
        }
        else if (data[0] == 0xBB)
        {
          int data_count = (size - 22) / 2;
          if (data_count < 0)
          {
            ROS_WARN("S3000: bad data count (%d)", data_count);
            memmove(rx_buffer, &rx_buffer[size+4], rx_count - (size+4));
            rx_count -= (size + 4);
            continue;
          }
          if (!recognisedScanner)
            SetScannerParams(scan, data_count); // Set up parameters based on number of results.

          // Scan data, clear ranges, keep configuration
          scan.ranges.clear();
          scan.intensities.clear();  // not used
          scan.ranges.resize(data_count);

          for (int ii = 0; ii < data_count; ++ii)
          {
            unsigned short Distance_CM = (*reinterpret_cast<unsigned short *> (&data[4 + 2*ii]));
            Distance_CM &= 0x1fff; // remove status bits
            double distance_m = static_cast<double>(Distance_CM)/100.0;

            if (mirror == 1)
              scan.ranges[data_count - ii - 1] = static_cast<float> (distance_m); // Reverse order.
            else
              scan.ranges[ii] = static_cast<float> (distance_m);
          }

          // CHECK
          // Return this flag to let the node know that the message is ready to publish
          bValidData = true;
        }
        else if (data[0] == 0xCC)
        {
          ROS_WARN("We got a reflector data packet we dont know what to do with it\n");
        }
        else
        {
          ROS_WARN("We got an unknown packet\n");
        }
      }
    }

    memmove(rx_buffer, &rx_buffer[size+4], rx_count - (size+4));
    rx_count -= (size + 4);
    continue;
  }
  return 1;
}


static const unsigned short crc_table[256] = {
  0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50a5, 0x60c6, 0x70e7,
  0x8108, 0x9129, 0xa14a, 0xb16b, 0xc18c, 0xd1ad, 0xe1ce, 0xf1ef,
  0x1231, 0x0210, 0x3273, 0x2252, 0x52b5, 0x4294, 0x72f7, 0x62d6,
  0x9339, 0x8318, 0xb37b, 0xa35a, 0xd3bd, 0xc39c, 0xf3ff, 0xe3de,
  0x2462, 0x3443, 0x0420, 0x1401, 0x64e6, 0x74c7, 0x44a4, 0x5485,
  0xa56a, 0xb54b, 0x8528, 0x9509, 0xe5ee, 0xf5cf, 0xc5ac, 0xd58d,
  0x3653, 0x2672, 0x1611, 0x0630, 0x76d7, 0x66f6, 0x5695, 0x46b4,
  0xb75b, 0xa77a, 0x9719, 0x8738, 0xf7df, 0xe7fe, 0xd79d, 0xc7bc,
  0x48c4, 0x58e5, 0x6886, 0x78a7, 0x0840, 0x1861, 0x2802, 0x3823,
  0xc9cc, 0xd9ed, 0xe98e, 0xf9af, 0x8948, 0x9969, 0xa90a, 0xb92b,
  0x5af5, 0x4ad4, 0x7ab7, 0x6a96, 0x1a71, 0x0a50, 0x3a33, 0x2a12,
  0xdbfd, 0xcbdc, 0xfbbf, 0xeb9e, 0x9b79, 0x8b58, 0xbb3b, 0xab1a,
  0x6ca6, 0x7c87, 0x4ce4, 0x5cc5, 0x2c22, 0x3c03, 0x0c60, 0x1c41,
  0xedae, 0xfd8f, 0xcdec, 0xddcd, 0xad2a, 0xbd0b, 0x8d68, 0x9d49,
  0x7e97, 0x6eb6, 0x5ed5, 0x4ef4, 0x3e13, 0x2e32, 0x1e51, 0x0e70,
  0xff9f, 0xefbe, 0xdfdd, 0xcffc, 0xbf1b, 0xaf3a, 0x9f59, 0x8f78,
  0x9188, 0x81a9, 0xb1ca, 0xa1eb, 0xd10c, 0xc12d, 0xf14e, 0xe16f,
  0x1080, 0x00a1, 0x30c2, 0x20e3, 0x5004, 0x4025, 0x7046, 0x6067,
  0x83b9, 0x9398, 0xa3fb, 0xb3da, 0xc33d, 0xd31c, 0xe37f, 0xf35e,
  0x02b1, 0x1290, 0x22f3, 0x32d2, 0x4235, 0x5214, 0x6277, 0x7256,
  0xb5ea, 0xa5cb, 0x95a8, 0x8589, 0xf56e, 0xe54f, 0xd52c, 0xc50d,
  0x34e2, 0x24c3, 0x14a0, 0x0481, 0x7466, 0x6447, 0x5424, 0x4405,
  0xa7db, 0xb7fa, 0x8799, 0x97b8, 0xe75f, 0xf77e, 0xc71d, 0xd73c,
  0x26d3, 0x36f2, 0x0691, 0x16b0, 0x6657, 0x7676, 0x4615, 0x5634,
  0xd94c, 0xc96d, 0xf90e, 0xe92f, 0x99c8, 0x89e9, 0xb98a, 0xa9ab,
  0x5844, 0x4865, 0x7806, 0x6827, 0x18c0, 0x08e1, 0x3882, 0x28a3,
  0xcb7d, 0xdb5c, 0xeb3f, 0xfb1e, 0x8bf9, 0x9bd8, 0xabbb, 0xbb9a,
  0x4a75, 0x5a54, 0x6a37, 0x7a16, 0x0af1, 0x1ad0, 0x2ab3, 0x3a92,
  0xfd2e, 0xed0f, 0xdd6c, 0xcd4d, 0xbdaa, 0xad8b, 0x9de8, 0x8dc9,
  0x7c26, 0x6c07, 0x5c64, 0x4c45, 0x3ca2, 0x2c83, 0x1ce0, 0x0cc1,
  0xef1f, 0xff3e, 0xcf5d, 0xdf7c, 0xaf9b, 0xbfba, 0x8fd9, 0x9ff8,
  0x6e17, 0x7e36, 0x4e55, 0x5e74, 0x2e93, 0x3eb2, 0x0ed1, 0x1ef0
};

unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
{
  unsigned short CRC_16 = 0xFFFF;
  unsigned short i;
  for (i = 0; i < length; i++)
  {
    CRC_16 = (CRC_16 << 8) ^ (crc_table[(CRC_16 >> 8) ^ (Data[i])]);
  }
  return CRC_16;
}


include/sicks3000.h
/*
*  Copyright (c) 2012, Robotnik Automation, SLL
*
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/
/*
 Desc: Driver for the SICK S3000 laser
 Author: Robotnik Automation SLL (based on sicks3000 by Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard for Player/Stage)
 Date: 1 Sept 2012

 The sicks3000 driver controls the SICK S 3000 safety laser scanner interpreting its data output.
 The driver is very basic and assumes the S3000 has already been configured to continuously output
 its measured data on the RS422 data lines.
*/
#include <assert.h>
#include <math.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <ros/ros.h>
#include "sensor_msgs/LaserScan.h"
#include "SerialDevice.h"

#define S3000_DEFAULT_TRANSFERRATE 500000
#define S3000_DEFAULT_PARITY	   "none"
#define S3000_DEFAULT_DATA_SIZE    8

// The laser device class.
class SickS3000
{
  public:

    // Constructor
    SickS3000(std::string port);

    // Destructor
    ~SickS3000();

    //! Open the port
    int Open();

    //! Close the port
    int Close();

    //! Interrupt continuous data from laser, returning whether interruption was successful.
    bool interruptDevice(int tries = 10);

    //! Read the operating data block and SCID of the S3000 for system and error status
    int getDiagnosticInfo(bool host, uint16_t* scid, uint8_t* seven_seg_first_char, uint8_t* seven_seg_second_char);

    //! Command the LIDAR network to restart.
    int triggerRestart();

    //! Get the current error cause and suggested action, if an error exists
    void GetDeviceError(uint8_t first_char, uint8_t second_char, std::string* error_code,
                        std::string* error, std::string* suggested_action);

    //! Wait for scan data to read - required call prior to ReadLaser!
    int WaitForScan(sensor_msgs::LaserScan& scan);

    //! Read and process data
    int ReadLaser(sensor_msgs::LaserScan& scan_msg, bool& bValidData); // public periodic function

  private:

    // Process range data from laser
    int ProcessLaserData(sensor_msgs::LaserScan& scan_msg, bool& bValidData); // public periodic function

    // Calculates CRC for a telegram
    unsigned short CreateCRC(uint8_t *data, ssize_t len);

    // Get the time (in ms)
    int64_t GetTime();

  protected:

    // serial port
    SerialDevice* serial;

    // Defines if laser is mounted inverted
    int mirror;

    // Scan width and resolution.
    int scan_width, scan_res;

    // Start and end scan angles (for restricted scan).  These are in
    // units of 0.01 degrees.
    int min_angle, max_angle;

    // Start and end scan segments (for restricted scan).  These are
    // the values used by the laser.
    int scan_min_segment, scan_max_segment;

    bool recognisedScanner;

    // rx buffer
    uint8_t * rx_buffer;
    unsigned int rx_buffer_size;
    unsigned int rx_count;
};


src/s3000_laser.cc
/*********************************************************************
*
*  This program uses the sicks3000  component to get laser scans, and then
*  publishes them as ROS messages
*
*  Copyright (c) 2012, Robotnik Automation, SLL
*
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*********************************************************************/

#include <assert.h>
#include <math.h>
#include <iostream>
#include <boost/format.hpp>
#include <boost/scoped_ptr.hpp>

#include "sicks3000.h"   // s3000 driver from player (Toby Collet / Andrew Howard)
#include "ros/time.h"
#include "diagnostic_updater/DiagnosticStatusWrapper.h"
#include "diagnostic_updater/diagnostic_updater.h"
#include "diagnostic_updater/publisher.h"
#include "diagnostic_updater/update_functions.h"

#include "sensor_msgs/LaserScan.h"
#include "std_srvs/Empty.h"

#include "std_msgs/Bool.h"

#define RADIANS(X) ((X) * 3.141592653589793 / 180.0)


using namespace std;

class s3000node {

public:
  boost::scoped_ptr<SickS3000> laser_;
  sensor_msgs::LaserScan scan_msg_;

  string port_;

  diagnostic_updater::Updater diagnostic_;

  ros::NodeHandle node_handle_;
  ros::NodeHandle private_node_handle_;

  enum {
    STATE_DISCONNECTED,
    STATE_CONNECTED,
    STATE_RESTARTING,
    STATE_ERROR_SCANS,
    STATE_OK
  } state_;

  bool host_lidar_;
  bool enable_scid_error_;
  int expected_scid_;
  uint16_t scid_;
  bool tx_failure_;
  uint8_t seven_seg_first_char_;
  uint8_t seven_seg_second_char_;
  ros::Time error_state_start_;

  /**
   * List of error codes which we are permitted to restart the LIDAR out of.
   */
  ros::V_string restart_errors_;

  /**
   * Times we have restarted without a 60 second period of normal operation
   * in between.
   */
  int successive_restarts_;

  /**
   * When this limit is hit, no more restarting. This prevents us from cycling
   * the LIDAR when it is doing no good.
   */
  int successive_restart_limit_;

  /**
   * Tracks errors detected for the lifetime of the node; the string is an
   * error_code value, and the int is a counter.
   */
  typedef std::map<std::string, int> ErrorCountersMap;
  ErrorCountersMap error_counters_;

  string frameid_;

  double desired_freq_;

  double range_max_;  // [m]
  double range_min_;  // [m]

  typedef diagnostic_updater::DiagnosedPublisher<sensor_msgs::LaserScan> LaserScanDiagnosedPublisher;
  boost::scoped_ptr<LaserScanDiagnosedPublisher> data_pub_;

  s3000node(ros::NodeHandle h) :
    node_handle_(h), private_node_handle_("~"),
    state_(STATE_DISCONNECTED),
    successive_restarts_(0),
    successive_restart_limit_(10),
    desired_freq_(16.6),
    range_max_(51.0),  // NOTE datasheet reports 52.0m, but actual is 51.2m
    range_min_(0.051)   // Determined by observation.
  {
    private_node_handle_.param("port", port_, string("/dev/ttyUSB0"));
    private_node_handle_.param("frame_id", frameid_, string("laser"));
    private_node_handle_.param("range_max", range_max_, range_max_);
    private_node_handle_.param("range_min", range_min_, range_min_);
    private_node_handle_.param("host_lidar", host_lidar_, true);
    private_node_handle_.param("expected_scid", expected_scid_, -1);
    private_node_handle_.param("enable_scid_error", enable_scid_error_, true);
    private_node_handle_.param("successive_restart_limit", successive_restart_limit_, successive_restart_limit_);

    if (private_node_handle_.hasParam("restart_errors"))
    {
      private_node_handle_.getParam("restart_errors", restart_errors_);
    }


    double desired_freq_tolerance = 0.2;
    private_node_handle_.param("desired_frequency", desired_freq_, desired_freq_);
    private_node_handle_.param("desired_frequency_tolerance", desired_freq_tolerance, desired_freq_tolerance);

    scan_msg_.header.frame_id = frameid_;
    scan_msg_.angle_min = static_cast<float>(RADIANS(-95.0));
    scan_msg_.angle_max = static_cast<float>(RADIANS(95.0));
    scan_msg_.angle_increment = static_cast<float>(RADIANS(0.25));
    scan_msg_.range_min = range_min_;
    scan_msg_.range_max = range_max_;

    diagnostic_.add("connection status", this, &s3000node::deviceStatusDiagnostics);

    diagnostic_.add("errors", this, &s3000node::errorLogDiagnostics);

    data_pub_.reset(new LaserScanDiagnosedPublisher(
          node_handle_.advertise<sensor_msgs::LaserScan>("scan", 1), diagnostic_,
          diagnostic_updater::FrequencyStatusParam(&desired_freq_, &desired_freq_, desired_freq_tolerance),
          diagnostic_updater::TimeStampStatusParam()));

    laser_.reset(new SickS3000(port_));

    scid_ = 0x0000;
    seven_seg_first_char_ = 0x00;
    seven_seg_second_char_ = 0x00;
    tx_failure_ = false;
  }


  ~s3000node()
  {
    stop();
  }


  bool start()
  {
    stop();

    if (laser_->Open() == 0)
    {
      diagnostic_.setHardwareID("Laser Ranger");
      ROS_INFO("Laser Ranger sensor initialized.");
      state_ = STATE_CONNECTED;
      return true;
    }
    else
    {
      ROS_ERROR_THROTTLE(10.0, "Laser Ranger sensor initialization failed.");
      ROS_DEBUG("Laser Ranger sensor initialization failed.");
      state_ = STATE_DISCONNECTED;
      diagnostic_.force_update();
      return false;
    }
  }


  void stop()
  {
    if (state_ >= STATE_CONNECTED)
    {
      laser_->Close();
      state_ = STATE_DISCONNECTED;
    }
  }


  void ConnectTest(diagnostic_updater::DiagnosticStatusWrapper& status)
  {
    if (laser_->Open() == 0)
    {
      status.summary(diagnostic_msgs::DiagnosticStatus::OK, "Connected successfully.");
    }
    else
    {
      status.summary(diagnostic_msgs::DiagnosticStatus::ERROR, "Connection unsuccessful.");
    }
  }

  bool spin()
  {
    // Outer loop for reconnection.
    while (ros::ok())
    {
      diagnostic_.update();
      ros::spinOnce();

      if (start())
      {
        // Inner loop runs once per received scan.
        while(ros::ok())
        {
          laser_->WaitForScan(scan_msg_);
          ros::Duration(0.04).sleep();

          bool scan_available = false;
          if (laser_->ReadLaser(scan_msg_, scan_available) < 0)
          {
            if (state_ != STATE_RESTARTING)
            {
              // Problem reading, or read timed out. Disconnect and try again.
              state_ = STATE_DISCONNECTED;
              ros::Duration(3.0).sleep();
              break;
            }
          }

          if (scan_available)
          {
            if (scan_msg_.ranges[0] > range_min_)
            {
              // Looks like a valid scan.
              state_ = STATE_OK;
              scan_msg_.scan_time = (1.0 / desired_freq_);
              scan_msg_.time_increment = scan_msg_.scan_time / scan_msg_.ranges.size();
              data_pub_->publish(scan_msg_);

              if (ros::Time::now() - error_state_start_ > ros::Duration(60.0) && successive_restarts_ > 0)
              {
                // We are receiving valid scans, and it has been 60 seconds since our
                // most recent error -> reset the counter.
                successive_restarts_ = 0;
                ROS_DEBUG("successive_restarts = %d", successive_restarts_);
              }
            }
            else
            {
              // When in an error state, the LIDAR reports all minimum range values.
              // Setting the state here flags the diagnostics function to query the LIDAR
              // for status code, and optionally reset it.
              ROS_DEBUG("LIDAR scan values indicate scanner error.");
              state_ = STATE_ERROR_SCANS;
            }
          }

          diagnostic_.update();
          ros::spinOnce();
        }
      }
      else
      {
        ROS_INFO("Retrying in 1 second.");
        ros::Duration(1.0).sleep();
      }
    }

    stop();

    return true;
  }

  void errorLogDiagnostics(diagnostic_updater::DiagnosticStatusWrapper &status)
  {
    int total = 0;

    ErrorCountersMap::iterator it;
    for (it = error_counters_.begin(); it != error_counters_.end(); ++it)
    {
      total += it->second;
      std::stringstream ss;
      ss << "Error count for code [" << it->first << "]";
      status.add(ss.str(), it->second);
    }

    if (total == 0)
    {
      status.summary(diagnostic_msgs::DiagnosticStatus::OK, "No LIDAR errors seen.");
    }
    else
    {
      status.summaryf(diagnostic_msgs::DiagnosticStatus::OK, "%d LIDAR error(s) seen.", total);
    }
  }

  void deviceStatusDiagnostics(diagnostic_updater::DiagnosticStatusWrapper &status)
  {
    switch (state_)
    {
      case STATE_DISCONNECTED:
        status.summary(diagnostic_msgs::DiagnosticStatus::ERROR, "LIDAR not connected.");
        break;
      case STATE_CONNECTED:
        status.summary(diagnostic_msgs::DiagnosticStatus::OK, "LIDAR connected, not receiving scans.");
        break;
      case STATE_RESTARTING:
        status.summary(diagnostic_msgs::DiagnosticStatus::OK, "LIDAR commanded to restart.");
        break;
      case STATE_ERROR_SCANS:
        // Delegate this case to a separate function.
        errorDiagnostics(status);
        break;
      case STATE_OK:
        status.summary(diagnostic_msgs::DiagnosticStatus::OK, "LIDAR OK.");
        if (scid_ == 0)
        {
          // If we just started up and don't yet know our SCID, trigger the error diagnostics one time
          // to find out.
          errorDiagnostics(status);
        }
        break;
    }


    if ((expected_scid_ != scid_) && (expected_scid_ != -1) && !tx_failure_)
    {
      status.summary(enable_scid_error_ ? (int)diagnostic_msgs::DiagnosticStatus::ERROR :
                                          (int)diagnostic_msgs::DiagnosticStatus::WARN,
                     "LIDAR configuration has invalid SCID.");
    }

    status.add("Device", port_);
    status.add("TF Frame", frameid_);
    status.add("Range max", range_max_);

    status.addf("SCID", "0x%04X", scid_);

    if (expected_scid_ != -1)
    {
      status.addf("Expected SCID", "0x%04X", expected_scid_);
    }
    else
    {
      status.add("Expected SCID", "Unset");
    }
  }

  void errorDiagnostics(diagnostic_updater::DiagnosticStatusWrapper &status)
  {
    bool first_time = false;

    if (tx_failure_)
    {
      // Early return when we've had a previous communications failure.
      status.summary(diagnostic_msgs::DiagnosticStatus::WARN,
          "LIDAR cannot be commanded; check comms interface.");
      return;
    }

    if (seven_seg_first_char_ == 0 && seven_seg_second_char_ == 0)
    {
      // Query for error code, since we don't know it yet. On successive calls,
      // assume it has not changed.
      first_time = true;
      if (laser_->getDiagnosticInfo(host_lidar_, &scid_,
            &seven_seg_first_char_, &seven_seg_second_char_) < 0)
      {
        // Unable to interact with the device. Mark us as having a communications
        // failure so that we don't keep on re-attempting it.
        tx_failure_ = true;
      }
    }

    // Only do the stuff below if there's an actual error. There might not be,
    // since we also trigger this check on device startup.
    if (seven_seg_first_char_ != 0 || seven_seg_second_char_ != 0)
    {
      std::string error_code;
      std::string error_description;
      std::string error_suggestion;
      laser_->GetDeviceError(seven_seg_first_char_, seven_seg_second_char_, &error_code,
                             &error_description, &error_suggestion);

      if (first_time)
      {
        error_state_start_ = ros::Time::now();
        successive_restarts_++;
        ROS_DEBUG("successive_restarts = %d", successive_restarts_);

        if (error_counters_.count(error_code) == 0)
        {
          error_counters_[error_code] = 0;
        }
        error_counters_[error_code]++;
      }

      status.add("Error Code", error_code);
      status.add("Error Description", error_description);
      status.add("Suggested Actions", error_suggestion);
      status.summaryf(diagnostic_msgs::DiagnosticStatus::WARN, "LIDAR error (%s)", error_code.c_str());

      // There is a race condition where we can sometimes end up catching the LIDAR
      // showing its startup code. This logic will suppress that and cause us to poll
      // its error code again next time into this function.
      if (error_code == "H" || error_code == "G" || error_code == "?" || error_code == "")
      {
        ROS_WARN("Suppressing error code in order to re-check.");
        seven_seg_first_char_ = 0;
        seven_seg_second_char_ = 0;
      }

      if (host_lidar_ && canRestartError(error_code) &&
          successive_restarts_ < successive_restart_limit_ &&
          (ros::Time::now() - error_state_start_ > ros::Duration(2.0)))
      {
        ROS_INFO("Triggering LIDAR restart to attempt to clear error.");
        laser_->triggerRestart();

        // Clear the stored error code so that we know to query it again if there's
        // another error within the lifespan of this node.
        seven_seg_first_char_ = 0;
        seven_seg_second_char_ = 0;
        state_ = STATE_RESTARTING;
      }
    }
  }

  bool canRestartError(std::string error_code)
  {
    for (int i = 0; i < restart_errors_.size(); i++)
    {
      if (error_code == restart_errors_[i])
      {
        return true;
      }
    }
    return false;
  }
};


int main(int argc, char** argv)
{
  ros::init(argc, argv, "s3000_node");

  ros::NodeHandle nh;

  s3000node s3000n(nh);
  s3000n.spin();

  return(0);
}
